lib/rucio/tests/test_api_external_representation.py

```diff
@@ -128,3 +161,257 @@ def test_api_account_limit(self):
+        add_replicas('MOCK', [{'scope': self.scope_name, 'name': did, 'bytes': 100}], issuer='root', **self.vo)
+
+        add_did(self.scope_name, 'ext_parent_2', 'dataset', issuer='root', account=self.account_name, **self.vo)
+        attachment = {'scope': self.scope_name, 'name': 'ext_parent_2',
+                      'dids': [{'scope': self.scope_name, 'name': did}]}
+        attach_dids_to_dids([attachment], issuer='root', **self.vo)
+
+        # get_did_from_pfns - in progress but requires other fixes
+        # protocol = Default(protocol_attr={'auth_token': None, 'prefix': '/tmp/rucio_rse/',
+        #                                   'scheme': 'https', 'hostname': 'mock.com', 'port': 8443},
+        #                    rse_settings={'rse': self.rse_name, 'id': self.rse_id, 'rse_id': self.rse_id, 'deterministic': False})
+        # lfns = {'scope': self.scope.internal, 'name': did} #should not need internal, this is because of protocol behaviour
+        # pfns = protocol.lfns2pfns(lfns)
+
+        # out = get_did_from_pfns(pfns, self.rse_name)
+        # for p in out:
+        #     assert_in('@', p)
+        #     for key in p:
+        #         assert_in('@', p[key]['scope'])
+
+        out = list_replicas(dids=[{'scope': self.scope_name, 'name': did}], resolve_parents=True)
+        for rep in out:
+            assert_equal(rep['scope'], self.scope_name)
+            if 'parents' in rep:
+                for parent in rep['parents']:
+                    assert_in(self.scope_name, parent)
+                    if self.multi_vo:
+                        assert_not_in(self.scope.internal, parent)
+
+    def test_api_request(self):
+        """ REQUEST (API): Test external representation of requests """
+
+        did = generate_uuid()
+        add_did(self.scope_name, did, 'dataset', issuer='root', account=self.account_name, rse=self.rse_name, **self.vo)
+
+        requests = [{
+            'dest_rse_id': self.rse2_id,
+            'source_rse_id': self.rse_id,
+            'request_type': constants.RequestType.TRANSFER,
+            'request_id': generate_uuid(),
+            'name': did,
+            'scope': self.scope_name,
+            'account': self.account_name,
+            'rule_id': generate_uuid(),
+            'retry_count': 1,
+            'requested_at': datetime.now(),
+            'attributes': {
+                'activity': 'User Subscription',
+                'bytes': 10,
+                'md5': '',
+                'adler32': ''
+            }
+        }]
+
+        reqs = queue_requests(requests, issuer='root', **self.vo)  # this does not pass in the source rse
+        for r in reqs:
+            assert_equal(r['scope'], self.scope_name)
+            assert_equal(r['account'], self.account_name)
+            assert_equal(r['source_rse'], self.rse_name)
+            assert_equal(r['dest_rse'], self.rse2_name)
+
+        # test get_next - difficult as will not necessarily return our queued requests.
```

Comment:
Just want to make sure none of these slip through the cracks, but if you already have a record of them/plan to work on them then ignore me.

Reply:
This one just requires a bit more thought. `get_next` returns the next request in the queue, but other requests generated by other tests may appear at the top of the queue. 

In the multi-VO case you could just put the requests on a different VO so you know they're top of the queue, but that doesn't work for single-VO.

I will probably uncomment the first four lines of the test as-is, if I don't get any better ideas.

Reply:
Actually, in this case, `get_next`'s output is passed through `api_update_return_dict`, so it's probably not necessary to test (I've not written tests for other things that use `api_update_return_dict`, I've just tested the function itself). I'll take this out.